### DISCUSSION TITLE
Can I resolve a cache item with a non-key field?

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2525] [Q&A] [answered]
This was originally discussed at https://github.com/FormidableLabs/urql/issues/762, but it's been 2 years so I thought I should open a new discussion in case something changed.

I want to query `Venue` by _either_ `id` or `slug`:

```graphql
query Venue($slug: String, $id: String) {
  venue(slug: $slug, id: $id) {
    id
    slug
    name
  }
}
```

Some places query by `slug`, such as a url for `/venues/some-slug`. Some query by ID. 

[This comment](https://github.com/FormidableLabs/urql/issues/762#issuecomment-620014260) says that it isn't possible to retrieve an item from the cache from a non-key field.

> It’s not possible to resolve items when the key can’t be generated. All items are normalized and indexed by the customizable key;

Is this still the case, or can I set up a resolver that gets the `id` from the `slug` somehow?

For example, it would be nice to do something like this, even when the venue key is determined by `venue.id`, and not `venue.slug`. (Pseudocode to illustrate).

```ts
// pseudocode in a Query resolver
const venue = cache.resolve({ __typename: 'Venue', slug: args.slug })
const id = venue?.id

return { __typename: 'Venue', id }
```

I tried the same with `readFragment`, but as expected, it didn't work.

Thanks again!

**Comment by JoviDeCroock** [comment, marked as answer]
https://github.com/FormidableLabs/urql/discussions/2519#discussioncomment-3015242 contains some pointers around this

**Reply by nandorojo** [reply]
Marking this as an answer since that looks like a duplicate.
